src/manifold/src/face_op.cpp

```diff
@@ -35,10 +40,130 @@ void Manifold::Impl::Face2Tri(const VecDH<int>& faceEdge,
-  triNormal.reserve(halfedge_.size());
-  triRef.reserve(halfedge_.size());
+#if MANIFOLD_PAR == 'T' && __has_include(<tbb/tbb.h>)
+  tbb::task_group group;
+  // map from face to triangle
+  tbb::concurrent_unordered_map<int, std::vector<glm::ivec3>> results;
+  VecDH<int> triCount(faceEdge.size());
+  triCount.back() = 0;
+  // precompute number of triangles per face, and launch async tasks to
+  // triangulate complex faces
+  for_each_host(
+      autoPolicy(faceEdge.size()), countAt(0), countAt(faceEdge.size() - 1),
+      [&](int face) {
+        triCount[face] = faceEdge[face + 1] - faceEdge[face] - 2;
+        ASSERT(triCount[face] >= 1, topologyErr,
+               "face has less than three edges.");
+        if (triCount[face] > 2)
+          group.run([&, face] {
+            const glm::vec3 normal = faceNormal_[face];
+            const glm::mat3x2 projection = GetAxisAlignedProjection(normal);
+            const PolygonsIdx polys = Face2Polygons(face, projection, faceEdge);
+            std::vector<glm::ivec3> newTris = TriangulateIdx(polys, precision_);
+            triCount[face] = newTris.size();
+            results[face] = std::move(newTris);
+          });
+      });
+  group.wait();
+  // prefix sum computation (assign unique index to each face) and preallocation
+  exclusive_scan(autoPolicy(triCount.size()), triCount.begin(), triCount.end(),
+                 triCount.begin(), 0);
+  triVerts.resize(triCount.back());
+  triNormal.resize(triCount.back());
+  triRef.resize(triCount.back());
 
+  // set triangles in parallel
+  for_each_host(
+      autoPolicy(faceEdge.size()), countAt(0), countAt(faceEdge.size() - 1),
+      [&](int face) {
+        const int firstEdge = faceEdge[face];
+        const int lastEdge = faceEdge[face + 1];
+        const int numEdge = lastEdge - firstEdge;
+        const glm::vec3 normal = faceNormal_[face];
+        if (numEdge == 3) {  // Single triangle
+          int mapping[3] = {halfedge_[firstEdge].startVert,
+                            halfedge_[firstEdge + 1].startVert,
+                            halfedge_[firstEdge + 2].startVert};
+          glm::ivec3 tri(halfedge_[firstEdge].startVert,
+                         halfedge_[firstEdge + 1].startVert,
+                         halfedge_[firstEdge + 2].startVert);
+          glm::ivec3 ends(halfedge_[firstEdge].endVert,
+                          halfedge_[firstEdge + 1].endVert,
+                          halfedge_[firstEdge + 2].endVert);
+          if (ends[0] == tri[2]) {
+            std::swap(tri[1], tri[2]);
+            std::swap(ends[1], ends[2]);
+          }
+          ASSERT(ends[0] == tri[1] && ends[1] == tri[2] && ends[2] == tri[0],
+                 topologyErr, "These 3 edges do not form a triangle!");
+
+          triVerts[triCount[face]] = tri;
+          triNormal[triCount[face]] = normal;
+          triRef[triCount[face]] = halfedgeRef[firstEdge];
+        } else if (numEdge == 4) {  // Pair of triangles
```

Comment:
Can we reuse the old code from below for this with functions? It looks pretty duplicate.

Reply:
yes we can do that